### PR TITLE
[MODEL] BERT based NER

### PR DESCRIPTION
**Pull request by bikestra**
## Description ##

Early implementation of BERT based NER built on top of @kenjewu 's work. Not ready to be merged in, but per discussion in #593 I wanted to share the current state so that @fierceX , @parry2403 and others can chime in.

Note that current code depends on `seqeval` package, which I am not sure we should take dependency on.

### Essentials ###
- [X] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [X] Implement BERT model and basic training loop
- [?] Reproduce BERT paper's results
- [X] Clean up code for training loop, add more logs
- [ ] Remove dependency on `seqeval`

## Comments ##
- The best Test F1 score I got on CoNLL 2003 English is only 90.02, with parameters I put as default.

### CODE REVIEW

**Comment by mli** [comment]
Job PR-612/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/1/index.html

**Comment by mli** [comment]
Job PR-612/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/2/index.html

**Comment by mli** [comment]
Job PR-612/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/3/index.html

File: scripts/named_entity_recognition/code/train_bert_ner.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        trainer = mx.gluon.Trainer(net.collect_params(), 'adam', optimizer_params)
+
+    # collect differentiable parameters
+    logging.info("Collect params...")
+    # do not apply weight decay on LayerNorm and bias terms
+    for _, v in net.collect_params('.*beta|.*gamma|.*bias').items():
+        v.wd_mult = 0.0
+    params = [p for p in net.collect_params().values() if p.grad_req != 'null']
+
+    def train(data_loader):
```
Comment (fierceX) [inline scripts/named_entity_recognition/code/train_bert_ner.py]: It may be better to only calculate the effective length of `out` and `tag_ids`.
This is the case in my implementation:
```python
max_valid_length = max(valid_length).asscalar()
out = out[:, 1:max_valid_length-1, :]
label = label[:, 1:max_valid_length-1]
loss_value = loss_function(out, label.astype('float32').as_in_context(ctx)).mean()
```

Comment (bikestra) [reply scripts/named_entity_recognition/code/train_bert_ner.py]: Do you mean it would be computationally more efficient, because the loss will be computed on smaller number of tags? If you are talking about correctness, I am providing sample weights to zero-out loss values on tokens which we don't have to predict: https://github.com/dmlc/gluon-nlp/pull/612/files/cf7cbc87e3cb2ed83cac6fa3e065579ee0a98540#diff-aa85770855ab2b6db7a5b29c232749efR138

**Comment by mli** [comment]
Job PR-612/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/1/index.html

**Comment by mli** [comment]
Job PR-612/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/6/index.html

File: scripts/named_entity_recognition/code/__init__.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# coding: utf-8
```
Comment (hankcs) [inline scripts/named_entity_recognition/code/__init__.py]: BTW, a module named like `code` will [hide the stdlib module code](https://stackoverflow.com/questions/20157925/pycharm3-0-1win7-when-i-debug-my-webpy-project-it-output-importerror-cannot-im), disabling the debugger of PyCharm.

**Comment by mli** [comment]
Job PR-612/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/7/index.html

**Comment by bikestra** [comment]
@hankcs Thanks for your comments! Most of the comments seem to be from @kenjewu 's change here https://github.com/dmlc/gluon-nlp/pull/466 . I created a new PR #612 based on #466 , but since #466 has been open for a while, the situation is becoming awkward. Also, this PR uses very few functions from #466 , just a handful of data processing utility functions.

Here are some options:
* Make this part of BERT scripts https://github.com/dmlc/gluon-nlp/tree/master/scripts/bert .
  * Copy & paste #466 's data processing utility functions to here
  * Or, make these utility functions part of https://github.com/dmlc/gluon-nlp/blob/master/src/gluonnlp/data/conll.py
* Review CR #466 and push it first

@szha Any thoughts?

**Comment by szha** [comment]
@bikestra let's go with the first option, treat the all utility functions required here as new functionality in review standard, and try to merge this PR first.

**Comment by bikestra** [comment]
@szha sounds good. I will proceed in that direction.

**Comment by bikestra** [comment]
I removed code from the previous named entity recognition code by @kenjewu , so we don't have irrelevant code anymore.

**Comment by bikestra** [comment]
@hankcs  that would be fun to have. I will add it :)

File: scripts/bert/index.rst [hunk 3]
```diff
@@ -461,3 +461,24 @@ To get the score of the dev data, you need to download the dev dataset (`dev-v2.
...
+
+    $ python3 train_ner.py \
+        --train-path ${DATA_DIR}/train.txt \
+        --dev-path ${DATA_DIR}/dev.txt \
+        --test-path ${DATA_DIR}/test.txt
+        --gpu 0 --learning-rate 1e-5 --dropout-prob 0.1 --num-epochs 100 --batch-size 8 \
+        --optimizer bertadam --bert-model bert_24_1024_16 \
+        --save-checkpoint-prefix ${MODEL_DIR}/large_bert --seed 13531
+
+This achieves Test F1 from `91.5` to `92.2`.
```
Comment (szha) [inline scripts/bert/index.rst]: @bikestra could you upload a training log for this to dmlc/web-data?

**Comment by eric-haibin-lin** [comment]
@bikestra the CI is recently updated. Would you mind syncing with master? Thanks!

**Comment by bikestra** [comment]
Will add the log, and address code style problems per http://ci.mxnet.io/blue/organizations/jenkins/gluon-nlp/detail/PR-612/12/pipeline . @szha suggested that providing command line interface may not be within the scope of this PR, so will wrap this PR without the feature.

**Comment by mli** [comment]
Job PR-612/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/1/index.html

**Comment by mli** [comment]
Job PR-612/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/2/index.html

**Comment by bikestra** [comment]
@eric-haibin-lin I really like the idea of adding some tests, but CoNLL datasets are not public. Popular datasets for NER are not public, but maybe we can add something like WikiNER https://github.com/dice-group/FOX/tree/master/input/Wikiner , which is CC-BY-30.

File: scripts/bert/ner_model.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    -------
+        All fields from `bert_model.BERTTaggingDataset._encode_as_input`,
+        as well as the prediction of the model.
+
+    """
+    for data in data_loader:
+        text_ids, token_types, valid_length, tag_ids, flag_nonnull_tag = \
+            [x.astype('float32').as_in_context(ctx) for x in data]
+
+        with ExitStack() as stack:
```
Comment (bikestra) [inline scripts/bert/ner_model.py]: I didn't realize `ExitStack()` is not available in Python 2. We absolutely want python2 support, right?

Comment (eric-haibin-lin) [reply scripts/bert/ner_model.py]: I think so, at least gluonnlp still supports for now. Is there any walk-around?

**Comment by mli** [comment]
Job PR-612/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/3/index.html

**Comment by mli** [comment]
Job PR-612/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/4/index.html

**Comment by mli** [comment]
Job PR-612/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/5/index.html

File: scripts/bert/ner_model.py (hunk 4 above)
Comment (eric-haibin-lin) [reply scripts/bert/ner_model.py]: I think we can keep it, for now, for this release. The `seqeval` package seems not available for py2, and I've documented in `index.rst` that this example requires py3.

**Comment by mli** [comment]
Job PR-612/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-612/7/index.html
